### DISCUSSION TITLE
Emit signal / notification when the virtual keyboard opens and closes

### DISCUSSION THREAD

**Discussion by dominiks** [godotengine/godot-proposals#8442] [GUI]
Working on a mobile app I encountered the issue that the virtual keyboard that opens when editing a `LineEdit` blocks the user's view of the UI. It would be easy to fix this by adding a spacer `Control` that takes the height of `DisplayServer.virtual_keyboard_get_height()` (` / get_window().content_scale_factor` actually) to move all content out of that zone but there is no way to be notified when the view opens. You can enable this spacer element when the `LineEdit` receives focus and hide it when the focus is lost but the virtual keyboard can open and close without losing focus of the control.

Therefore I think it would be good to have a signal or a notification whenever the virtual keyboard is opened or closed. Currently the `DisplayServer` handles everything regarding the virtual keyboard but servers don't send signals, I think. Alternatively the event could be emitted by `LineEdit` and `TextEdit` as they automatically open the keyboard.

**Comment by duongxinh2003** [comment]
Any update?

**Reply by Calinou** [reply]
To my knowledge, nobody has started working on this yet. It should be possible but I'm not sure if the LineEdit/TextEdit sending the signal will lead to good usability.

Out of curioisity, what are you trying to achieve by having this signal available?

**Reply by roonskipp** [reply]
I would like this signal so I can resize a container to make sure part of my UI is visible right above the keyboard. 
I now do a check when I handle user input, but it feels unecessary to resize the container on every input change. I would rather do it on a signal emitted when the virtual keyboard is shown / hidden.

**Reply by JoNax97** [reply]
Would it make sense to implement it as an input event?

**Reply by Calinou** [reply]
It feels strange to have an input event type for a single set of actions (virtual keyboard open/close). A signal seems more appropriate to me.

**Reply by JoNax97** [reply]
Fair point. Could it be put in some global place? It would be weird to have it be a signal of an object instance.

**Comment by andris1** [comment]
Just want to add that I'm in a very similar position to @dominiks' - I'd like to move a bunch of UI above the virtual keyboard when it's open.

**Comment by Satellaa** [comment]
Any update guys?

**Comment by JoNax97** [comment]
I don't think anyone has gotten around to implementing this

**Comment by Gromph** [comment]
We have implemented this in a custom internal java android plugin.  It's got some hacks in it that may not be needed anymore, also I copied and pasted this out of a larger plugin, so it may not be a complete example, but here is how we did it:


```java
public class MonitorKeyboardAndroid extends GodotPlugin {
	private static final String TAG = "MonitorKeyboardAndroid ";
	int mVirtualKeyboardHeight;
	int mVirtualKeyboardHeightError;
	boolean mKeyboardShownOnce = false;

	@Nullable
	public View onMainCreate(Activity activity) {
		// listen for layout changes which generally indicates keyboard being shown/hidden and emit signals
		getGodot().getRenderView().getView().getViewTreeObserver().addOnGlobalLayoutListener(() -> {

			// Calculate keyboard height, this is copied from Godot.java::onVideoInit
			// It however is wrong, see https://github.com/godotengine/godot/issues/41388
			// Apparently it doesn't add the height of the android navigation bar
			// on my Pixel 5a with Android 12 the nav bar is a black bar with a down arrow to hide keyboard
			// When the keyboard is hidden keyboardHeight will be negative, it happens that this is the height 
			// of the nav bar, so adding it to keyboardHeight fixes it. Hopefully we can remove this when 41388 is fixed
			Point fullSize = new Point();
			getGodot().getActivity().getWindowManager().getDefaultDisplay().getSize(fullSize);
			Rect gameSize = new Rect();
			getGodot().getRenderView().getView().getWindowVisibleDisplayFrame(gameSize);
			final int keyboardHeight = fullSize.y - gameSize.bottom;

			if(keyboardHeight <= 0) {
				mVirtualKeyboardHeightError = keyboardHeight;
				mVirtualKeyboardHeight = 0;
			}

			if(android.os.Build.MANUFACTURER.equals("Amazon") && isTV() && keyboardHeight > fullSize.y * 0.75)
			{
				// Amazon Fire TV will trigger this callback when the keyboard is being hidden with mVirtualKeyboardHeight between 1027 and 1080
				// It causes our boxes to resizes rreally small and then large again causing a flicker, so just ignore it
				//Log.w(TAG, String.format("MonitorKeyboardAndroid::addOnGlobalLayoutListener ignoring large keyboard on Amazon Fire TV"));
			}
			else if(keyboardHeight > 0) {
				mVirtualKeyboardHeight = keyboardHeight + Math.abs(mVirtualKeyboardHeightError);
				mKeyboardShownOnce = true;
				emitSignal("keyboard_shown");
			}
			else {
				// only emit hidden signal if keyboard has been shown at least once. This hopefully fixes exception starting game where it
				// immediately tried to emit signal before signal was registered
				if(mKeyboardShownOnce)
					emitSignal("keyboard_hidden");
			}
		});
	}
	public boolean isTV() {
		PackageManager packageManager = getGodot().getActivity().getPackageManager();
		return packageManager.hasSystemFeature(PackageManager.FEATURE_LEANBACK);
	}
	@NonNull
	@Override
	public String getPluginName() {
		return "MonitorKeyboardAndroid ";
	}

	@NonNull
	@Override
	public Set<SignalInfo> getPluginSignals() {
		Set<SignalInfo> signals = new ArraySet<>();
		signals.add(new SignalInfo("keyboard_shown"));
		signals.add(new SignalInfo("keyboard_hidden"));
		return signals;
	}
}
```

**Comment by syntaxerror247** [comment]
It seems like a valid request to me. Before moving to implementation, we will need to discuss how this signal will be emitted. LineEdit and TextEdit seems good to me. We can also have a method in Mainloop, but I'm not sure about that.

**Comment by theraot** [comment]
These are the places it could be notified (that I can think of):

- Static signal: This is - [officially](https://github.com/godotengine/godot-proposals/issues/6851#issuecomment-1733027330) - [no supported](https://github.com/godotengine/godot-proposals/issues/6851).
- Signal on a singleton: The best logical fit would be `Input`, but kind of awkward. Or it can be on `DisplayServer` which is what actually controls it.
- Signal on some other object: It could be a signal on `Window`; or on `LineEdit` and `TextEdit`.
- Notification: I believe adding notifications is to be avoided... If it is common it is a problem, and if it isn't it is not worth it.
- InputEvent: I think this is fine, although I agree it might not be what you'd expect.

Now, I know people probably won't look in `DisplayServer` for the signal, but that seems to be the easier place to add it, and also it would be a central place to connect to instead of having it in both `LineEdit` and `TextEdit` and needing an instance of either to connect.

*By the way, note that `RenderingServer` already has signals. So a server with signals is on the table.*

Perhaps, for ease of use it can be surfaced on `Window`. Then we can easily get a hold of it with `get_window()` and connect to it from anywhere on the scene tree.

Now, I do not know if people are interested in reacting differently depending on which `LineEdit` or `TextEdit` (or what `Window` for that matter) requested the virtual keyboard. If that is what people want, then it makes sense to have the signal at that scope.

**Comment by stayinsidegames** [comment]
I'd be interested in this as well, although I don't have the skills to help implement it. My use case is this: When the virtual keyboard appears, the viewport resizes to be smaller because now the keyboard has taken some space (not the size of the whole keyboard, oddly, just perhaps a line of it at the bottom). As far as I can tell, there's no way to prevent the viewport from resizing, or to force it to retain its size, except perhaps turning scaling/stretching off, which is nonviable for a mobile app. When the viewport resizes, all elements in the screen shrink a little bit to accommodate the new, smaller screen size. When the keyboard disappears, all elements expand back to their old sizes. This contraction/expansion of elements is undesired, but I can't find a way around it.

Since I can't stop the viewport from resizing when the keyboard appears/disappears, having signals for when the keyboard opens/closes (preferably both begins/completes opening/closing) would help me make a hacky workaround, having the graphical elements fade out/in to avoid the jitter.

**Comment by skysphr** [comment]
`DisplayServer` already fires the signal `orientation_changed`, which is somewhat related to this, and I would consider it logical for a `virtual_keyboard_changed` to be placed under the same class, especially since it is the same place `virtual_keyboard_get_height()` is found. I find it strange for `LineEdit` or `TextEdit` objects to emit this signal, as I believe some systems have the ability to open or close the virtual keyboard at will, without the explicit request of the running application.
